src/binary_to_text/binary_encoder.ml

```diff
@@ -0,0 +1,713 @@
+  | Ref_is_null -> add_char '\xD1' buf
+  | Ref_func idx -> add_char_indice '\xD2' idx buf
+  | I_trunc_sat_f (S32, S32, S) -> fc_encode 0 buf
+  | I_trunc_sat_f (S32, S32, U) -> fc_encode 1 buf
+  | I_trunc_sat_f (S32, S64, S) -> fc_encode 2 buf
+  | I_trunc_sat_f (S32, S64, U) -> fc_encode 3 buf
+  | I_trunc_sat_f (S64, S32, S) -> fc_encode 4 buf
+  | I_trunc_sat_f (S64, S32, U) -> fc_encode 5 buf
+  | I_trunc_sat_f (S64, S64, S) -> fc_encode 6 buf
+  | I_trunc_sat_f (S64, S64, U) -> fc_encode 7 buf
+  | Memory_init idx ->
+    fc_encode 8 buf;
+    indice_encode idx buf;
+    add_char '\x00' buf
+  | Data_drop idx ->
+    fc_encode 9 buf;
+    indice_encode idx buf
+  | Memory_copy ->
+    fc_encode 10 buf;
+    add_char '\x00' buf;
+    add_char '\x00' buf
+  | Memory_fill ->
+    fc_encode 11 buf;
+    add_char '\x00' buf
+  | Table_init (tableidx, elemidx) ->
+    fc_encode 12 buf;
+    indice_encode elemidx buf;
+    indice_encode tableidx buf
+  | Elem_drop idx ->
+    fc_encode 13 buf;
+    indice_encode idx buf
+  | Table_copy (idx1, idx2) ->
+    fc_encode 14 buf;
+    indice_encode idx1 buf;
+    indice_encode idx2 buf
+  | Table_grow idx ->
+    fc_encode 15 buf;
+    indice_encode idx buf
+  | Table_size idx ->
+    fc_encode 16 buf;
+    indice_encode idx buf
+  | Table_fill idx ->
+    fc_encode 17 buf;
+    indice_encode idx buf
+  | _ -> assert false (* TODO *)
+
+and expr_encode expr ?end_op_code buf =
+  List.iter (fun instr -> instr_encode instr buf) expr;
+  match end_op_code with
+  | None -> add_char end_opcode buf
+  | Some end_op_code -> add_char end_op_code buf
+
+let export_encode cid (export : Binary.export) buf =
+  let { name; id } = export in
+  add_bytes name buf;
+  add_char cid buf;
+  write_U32_of_int id buf
+
+let global_encode (global : global) buf =
+  let { typ; init; id = _ } : global = global in
+  global_type_encode typ buf;
+  expr_encode init ?end_op_code:None buf
+
+let global_import_encode (global : _ global_type Imported.t) buf =
+  let { Imported.modul; name; assigned_name = _; desc } = global in
+  let mut, valtype = desc in
+  add_bytes modul buf;
+  add_bytes name buf;
+  add_char '\x03' buf;
+  valtype_encode valtype buf;
+  mut_encode mut buf
+
+let local_encode idx (local : _ param) buf =
+  let _so, vt = local in
+  write_U32_of_int idx buf;
+  valtype_encode vt buf
+
+let locals_encode typt_f locals buf =
+  let len = List.length locals in
+  let nb_params =
+    match typt_f with
+    | Bt_raw (_ind, (pt, _rt)) -> List.length pt
+    | _ -> assert false
+  in
+  write_U32_of_int len buf;
+  List.iteri (fun i loc -> local_encode (i + nb_params) loc buf) locals
+
+let encode_from_indexed func_local_encode func_import_encode ~is_local data buf
+    =
+  let v = Indexed.get data in
+  match v with
+  | Runtime.Local data ->
+    if is_local then Some (func_local_encode data buf) else None
+  | Runtime.Imported data ->
+    if is_local then None else Some (func_import_encode data buf)
```

Comment:
I think you should rather provide:

```ocaml
let keep_local =
  List.filter_map (fun data ->
    match Indexed.get data with
    | Local data -> Some data
    | Imported _data -> None )
```

And `keep_imported` too.

Then, in your `encode` function, you would do something like:

```ocaml
  let local_fun = keep_local modul.func in
  let imported_fun = keep_imported modul.func in
  ...

  section_encode '\x01' types_encode local_fun buf;
  section_encode '\x02' imports_encode
    (imported_func, imported_table, imported_mem, imported_global)
    buf;
  section_encode '\x03' functions_encode local_fun buf;
  ...
```

Reply:
Good review! Big refacto that has greatly simplified the code.